Interface changes to enable per channel quant.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* #37626 Perf optimization for conv and gemm kernels.
* #37624 Added per channel separate test cases for fc and deconv tests.
* #37623 Changes to enable per channel support on dynamic linear.
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* #37620 Changes to enable per channel requant.
* #37619 Enable per channel zero point.
* **#37618 Interface changes to enable per channel quant.**

Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

kimishpatel committed:
Interface changes to enable per channel quant.

Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -574,22 +560,49 @@ class QConvInt8 final {
+      uint8_t* weight_zp_data =
+        (uint8_t*)pack_data.w_zero_points.data_ptr<c10::quint8>();
+      float* weight_scales_data = pack_data.w_scales.data_ptr<float>();
+      pack_data.requantization_scale =
```
Maybe add a comment here about why we calculate requant scale here?

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -574,22 +560,49 @@ class QConvInt8 final {
+          weight_scales_data[0] * input_scale,
+          0, kQInt32);
+
+      qnnpack::conv_param_t conv_p(
```
Where is this used?

kimishpatel replied:
In PrePackConvWeights.

supriyar replied:
We create another instance of the param struct to pass to the run function. Can we avoid that?

kimishpatel replied:
Yes. Let me see why it is done twice. I dont quite recall now, but will try to get rid of one if it is not there for a reason, else will add a comment.

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -620,16 +633,27 @@ class QConvInt8 final {
+        groups,
+        C,
+        M,
+        (uint8_t*)pack_data.w_zero_points.data_ptr<c10::quint8>(),
```
please avoid using c style casts

kimishpatel replied:
This is from the original code.

kimishpatel replied:
But besides that do you mean to suggest use static_cast<uint8_t*>?

supriyar replied:
I think we use reinterpret_cast when accessing tensor data in most places.

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -574,22 +560,49 @@ class QConvInt8 final {
+          C,
+          M,
+          weight_zp_data,
+          pack_data.requantization_scale.data(),
```
Why aren't we passing `data_ptr`?

kimishpatel replied:
This is stored using std::vector and not tensor.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -93,11 +93,14 @@ struct conv_param_t {
+    //if (kernel_scale <= 0.0f || !std::isnormal(kernel_scale)) {
+    // This has to be checked for all the output channel requant scales.
+    // TODO
+    if (requantization_scales[0] <= 0.0f || !std::isnormal(requantization_scales[0])) {
```
I think we can check this when we create the requant scale

kimishpatel replied:
I think this is better here because the constraints come from qnnpack backend.

supriyar replied:
It would be nice if the calculation of requantization scale also happened on the backend side like before. Did you move it into the op so that we calculate it just once?

kimishpatel replied:
I think earlier it was calculated during requantization param creation. It we keep it there we have to populate that array every time we run the op. It is little less efficient. Plus it does seem natural that it is captured as part of the conv_params.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
+  float* weight_scales_data = weight_scales.data_ptr<float>();
+  std::vector<float> requant_scales(num_output_channels, 1.f);
+  for (int i = 0; i < num_output_channels; ++i) {
+    requant_scales[i] = weight_scales_data[i] * input_scale / output_scale;
```
Could we check requant_scale values here rather than in the run function?

kimishpatel replied:
We check the constraints inside qnnpack. Is that what you mean? Not sure what you mean by inside run function. This function is called from run function and values are checked during conv_param_t creation.

kimishpatel replied:
I will try to move the check here if it seems more natural.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
+  auto num_output_channels = weight_contig.size(0);
+  const auto qtype = weight_contig.qscheme();
+  at::Tensor weight_zp =
+    at::native::empty_cpu({1}, at::device(at::kCPU).dtype(at::kQUInt8));
```
Why quint8? fbgemm prepack step seems to be using int32_t.
Also why call empty_cpu explicitly here? Wouldn't empty() work?

kimishpatel replied:
Because qnnpack uses uint8 to store zero points. For empty_cpu it is to by pass all the dispatch overhead.

supriyar replied:
Is there a lot of overhead? Reason I ask is that we are still using aten::empty in other parts of the quantized code which is also run on mobile.

kimishpatel replied:
Well, in FBNet I saw this as an issue and calling directly the native function was getting rid of some overhead. Now it depends on the network. If the ops are taking much less time than such overhead matters.

kimishpatel replied:
Actually in this case, it should be much of overhead, because this is done only once.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
+  // Adjust weight zero point, similar to weight data.
+  uint8_t* weight_zp_data;
+  if (qtype == at::kPerTensorAffine) {
+    weight_zp_data = (uint8_t*)weight_zp.data_ptr<c10::quint8>();
```
Prefer reinterpret_cast over c style cast

kimishpatel replied:
Sure, but any particular reason or stylistic? I can make the change.

supriyar replied:
I think it is better to be explicit in the type of cast we are using. Also general convention in these files.

kimishpatel replied:
That sounds good.

supriyar replied:
I think it would be good to address the c style cast used everywhere in this PR.

kimishpatel replied:
Definitely. Will do that.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
+  uint8_t* weight_zp_data;
+  if (qtype == at::kPerTensorAffine) {
+    weight_zp_data = (uint8_t*)weight_zp.data_ptr<c10::quint8>();
+    weight_zp_data[0] = (uint8_t)(weight_contig.q_zero_point() + 128);
```
`q_zero_point()` returns int64_t. Why are we directly casting it to unit8_t here?

kimishpatel replied:
Are we expecting zero point to be outside of uint8_t range after the addition of 128? My understanding was the it is already representable in 8 bits. I would assume you would want to go via 64->32->8 for rounding reason but I dont know what difference it makes here.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
+    weight_zp_data[0] = (uint8_t)(weight_contig.q_zero_point() + 128);
+  } else if (qtype == at::kPerChannelAffine) {
+    weight_zp.resize_({num_output_channels});
+    weight_zp_data = (uint8_t*)weight_zp.data_ptr<c10::quint8>();
```
same here

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -574,22 +560,49 @@ class QConvInt8 final {
           reinterpret_cast<int8_t*>(weight_contig.data_ptr<c10::qint8>());
+
+      std::tie(pack_data.w_zero_points, pack_data.w_scales) =
+          make_zero_points_and_scales_tensor(weight_contig);
```
Actually this can be done in the conv_prepack function, right?

kimishpatel replied:
Yes but it is better for readability that it is here, otherwise you will have to go back and forth to understand what is happening.

supriyar replied:
I think as long as we add a comment it should suffice. We can get `pack_data.w_scale` as before, except that it will be a tensor of scales now.

kimishpatel replied:
My argument is that, we should not fill up pack_data, half in one place, half in another unless there is a good reason. It does not read well.

kimishpatel replied:
I think I am gonna have to this with torchbind support landed. But I have another way of doing per channel quant that will not require requant scales to be stored in this data structure. Hopefully then it will be cleaner as well.

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -305,11 +313,9 @@ class QLinearInt8 final {
-        input_contig.q_scale(),
-        kernel_zp,
-        kernel_scale,
+        (uint8_t*)pack_ptr.w_zero_points.data_ptr<c10::quint8>(),
```
Could you pass in weight_zp_data instead of re-calculating this here?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -29,43 +29,46 @@ static inline size_t compute_output_dimension(
+  uint8_t output_min;
+  uint8_t output_max;
+  bool transpose;
+  bool is_initialized;
```
Why remove the const qualifiers here?

kimishpatel replied:
So with const qualifiers I could not construct conv_param_t like `conv_p = conv_param_t(...` since it generates rvalues ref and default move assign cannot be used on consts.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/convolution.c:
```diff
@@ -475,11 +475,16 @@ enum pytorch_qnnp_status pytorch_qnnp_create_convolution2d_nhwc_q8(
         pytorch_qnnp_compute_requantization_params(
             convolution_scale, output_zero_point, output_min, output_max);
   } else {
+    // Passing address of kernel_zero_scale and convolution_scale
```
nit comment says kernel_zero_scale

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/fc-prepack.cc:
```diff
@@ -9,14 +9,14 @@ PackBMatrix::PackBMatrix(
-  if (kernel_scale <= 0.0f || !std::isnormal(kernel_scale)) {
+  if (requantization_scale <= 0.0f || !std::isnormal(requantization_scale)) {
     pytorch_qnnp_log_error(
         "failed to create fully connected operator with %.7g kernel scale: "
```
Shouldn't this say requantization_scale instead?

kimishpatel replied:
Aaah correct. My bad. Will fix.

supriyar reviewed (approved):
Left few comments. Please address them before landing

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -221,3 +227,69 @@ Tensor qnnpack_avg_pool2d(
+  float* weight_scales_data = weight_scales.data_ptr<float>();
+  std::vector<float> requant_scales(num_output_channels_padded, 1.f);
+  for (int i = 0; i < num_output_channels_padded; ++i) {
+    requant_scales[i] = weight_scales_data[i] * input_scale / output_scale;
```
For aligning numerics with FBGEMM, we want to compute the reciprocal of the output scale first:
inv_C_scale_fp32 = 1.0f / C_scale_fp32.
https://fb.quip.com/nsJmAabU2Bas

kimishpatel replied:
@raghuramank100, sure. Sorry I missed that. But that would have to be done in a separate PR.

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -29,21 +32,22 @@ struct PackedLinearWeightsQnnp : public LinearPackedParamsBase {
   c10::optional<double> input_scale;
-  double w_scale;
-  int64_t w_zp;
+  at::Tensor w_scales;
+  at::Tensor w_zero_points;
```
Why take these as at::Tensor rather than std::vector?

kimishpatel replied:
I think this can be made std::vector. Let me check.

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -221,3 +227,69 @@ Tensor qnnpack_avg_pool2d(
+    weight_zp_data = (uint8_t*)weight_zp.data_ptr<c10::quint8>();
+    weight_zp_data[0] = (uint8_t)(weight_contig.q_zero_point() + 128);
+  } else if (qtype == at::kPerChannelAffine) {
+    weight_zp.resize_({num_output_channels_padded});
```
Maybe zero-init the values after `num_output_channels` just to make this deterministic.

kimishpatel replied:
Sounds good.

dreiss reviewed (changes requested):
Looks like some TODOs need to be addressed.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -29,43 +29,46 @@ static inline size_t compute_output_dimension(
   size_t group_input_channels;
   size_t group_output_channels;
 
+  conv_param_t() : is_initialized(false) {}
+
```
I second this.  It's generally considered better practice if construction and initialization are not separated to avoid having an object in an invalid state.  I assume you are doing this because you need the default constructor, in which case the [admittedly painful] solution would be to overload the constructor of all objects that rely on an instance of conv_param_t to be default initialized if possible.

kimishpatel replied:
The "painful" solution is not quite possible in this scenario because we dont have all the information we need to construct this class at the time of its default instantiation. Unfortunately this is the design we have for QNNPACK integration of conv op. Let me think if we can change that.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:
```diff
@@ -497,7 +494,10 @@ enum pytorch_qnnp_status qnnpackConv(
+          // XZP kernels are not supporting per channel quant.
+          // We dont really use XZP kernels ATM.
+          // Thus assigning the zero point of first channel.
+          .multiplier = (int32_t)-conv_p.kernel_zero_points[0],
```
I don't think this cast is needed here.  Since an int32_t can represent any value that a uint8_t can, I think you should be OK without the cast.

kimishpatel replied:
I was just moving code around. Did not put the cast there myself. And I think it existed here by copying from the original operator-run.cc. So I am gonna leave it here.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/requantization.h:
```diff
@@ -159,12 +159,12 @@ pytorch_qnnp_compute_conv_quantization_params(
   for (uint32_t i = 0; i < 8; i++) {
     params.sse2.input_zero_point[i] = (int16_t)(uint16_t)input_zero_point;
-    params.sse2.kernel_zero_point[i] = (int16_t)(uint16_t)kernel_zero_point;
+    params.sse2.kernel_zero_point[i] = (int16_t)(uint16_t)kernel_zero_points[0];
```
Again I think these casts are not necessary.

kimishpatel replied:
Again all these exist here from earlier. Dont think these set of PRs is the right ones to fix it.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/convolution.c:
```diff
@@ -475,11 +475,16 @@ enum pytorch_qnnp_status pytorch_qnnp_create_convolution2d_nhwc_q8(
+    // is not safe as they are stack variables. However for now just staging this change.
+    // Eventually, we will find a better way to handle this.
+    // For now we dont go through this path for convolutions.
+    // TODO Kimish
```
If I understand this code correctly I think this usage is perfectly fine since you are dealing with a blocking function.  Unless you hold onto these pointers internally after this function returns and use them subsequently, which is indeed unsafe behavior, there is no way the variables these pointers point to can go out of scope before the function exits.

kimishpatel replied:
Yes so usage is fine. I just didnt feel comfortable passing an address of the stack variable. But you are right. It is safe otherwise.

kimishpatel replied:
Let me remove this comment.

kimishpatel commented:
> Looks like some TODOs need to be addressed.

@dreiss so the todos I put is for the entire stack. Because the way I have split stuff I put todo to make sure that in subsequent ones I fix them. So by the end of the stack there shouldn't be any todos that I put.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-prepack.cc:
```diff
@@ -142,9 +142,11 @@ PrePackConvWeights::PrePackConvWeights(
             packed_group_weights_size * groups);
         assert("QNNPACK Runtime Error.");
       }
+      // Setting packed weights to first channel's zero point.
+      // But this will be overwritten by weight packing.
```
So this is needed in this PR because some kernels explicitly rely on this. Otherwise you get wrong result. The comment is referring to the fact that later PRs will overwrite this. Let me fix this comment. This is not yet removed in the later PRs, I realize. So I will fix this in the appropriate PR.

kimishpatel commented:
> If you're only doing this once, why not check it for every output channel scale? It's just a few hundred checks at most, right?

@dreiss, this is done in the later PRs. I didn't do it in this PR because they are all supposed to be the same.

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]

kimishpatel committed:
Update on "Interface changes to enable per channel quant."


Summary:
This does not do any actual changes. Just introduces API changes and
some data struct changes to hold vector of data for zero point and
scale.
The idea is to pass a vector of zero points and requantization scales to QNNPACK backend. This is done even if we are doing per tensor quantization for weights. The reason as seen in subsequent diffs is to have one set of kernels do both per channel and per tensor quantization aware compute.

Test Plan:
Via unittests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.
PT's quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339039](https://our.internmc.facebook.com/intern/diff/D21339039)

[ghstack-poisoned]
